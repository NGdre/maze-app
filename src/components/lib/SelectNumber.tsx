import "./select-number.css";
import { useState, useId, FC, useEffect, ChangeEvent, FocusEvent } from "react";
import React from "react";
import { clamp } from "../../utils";

const classNames = {
  button:
    "bg-gray-300 text-gray-600 rounded-full outline-none cursor-pointer hover:text-white hover:bg-blue-500",
  buttonSpan: "m-auto text-2xl",
  selectNumberInput: `outline-none focus:outline-none text-center font-semibold 
              hover:text-gray-500 focus:text-black md:text-base flex items-center`,
  selectNumberLabel: "text-gray-700 text-sm font-semibold capitalize",
};

//@ts-ignore
classNames.buttonDisabled = classNames.button + " cursor-not-allowed";

// что-то не так со стилями
const UpdateCountButton = ({
  onClick,
  text,
  disabled,
  ...props
}: {
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  text: string;
  disabled: boolean;
}) => {
  return (
    <button
      {...props} // где правильней разместить props?
      onClick={onClick}
      disabled={disabled}
      //@ts-ignore
      className={disabled ? classNames.buttonDisabled : classNames.button}
    >
      {/* нельзя обойтись без span? */}
      <span className={classNames.buttonSpan}>{text}</span>
    </button>
  );
};

export type NumericValue = string | number;

interface SelectNumberProps {
  onSelect: (number: number) => void;
  value?: NumericValue;
  labelContent?: string;
  min?: NumericValue;
  max?: NumericValue;
}

const INCREASE_BTN_TEXT = "+";
const DECREASE_BTN_TEXT = "-";
const STEP = 1;

// компонент раздутый, наверняка с ним что-то не так
const SelectNumber: FC<SelectNumberProps> = ({
  labelContent = "",
  value = 0,
  onSelect,
  min = 0,
  max = 10,
  ...props
}) => {
  const [counter, updateCounter] = useState<NumericValue>(+value);
  const [prevValue, setPrevValue] = useState(value);
  const inputId = useId();

  min = +min;
  max = +max;

  const counterAsNumber = +counter;

  if (prevValue !== value) {
    if (value) {
      updateCounter(value);
      setPrevValue(+value);
    }
  }

  // I don't know how to do this without useEffect
  useEffect(() => {
    const maybeNumber = parseInt(String(counter));

    if (Number.isNaN(maybeNumber)) return;

    if (maybeNumber <= max && maybeNumber >= min) onSelect(maybeNumber);
  }, [counter]);

  const handleDecrement = () => {
    const incrementedCounter = counterAsNumber - STEP;

    if (incrementedCounter >= min) {
      updateCounter(incrementedCounter);
    }
  };

  const handleIncrement = () => {
    const decrementedCounter = counterAsNumber + STEP;

    if (decrementedCounter <= max) {
      updateCounter(decrementedCounter);
    }
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target;

    updateCounter(value);
  };

  const handleBlur = (event: FocusEvent<HTMLInputElement>) => {
    const { value } = event.target;

    let maybeNumber = parseInt(value, 10);

    if (Number.isNaN(maybeNumber)) {
      maybeNumber = min;
    }

    const clamped = clamp(maybeNumber, min, max);

    updateCounter(clamped);
  };

  return (
    <div className="grid grid-cols-3">
      <div className="text-center col-span-3">
        <label htmlFor={inputId} className={classNames.selectNumberLabel}>
          {labelContent}
        </label>
      </div>
      <UpdateCountButton
        onClick={handleDecrement}
        text={DECREASE_BTN_TEXT}
        disabled={counterAsNumber <= min}
      />
      <input
        {...props}
        type="number"
        id={inputId}
        value={counter}
        className={classNames.selectNumberInput}
        onChange={handleChange}
        onBlur={handleBlur}
        min={min}
        max={max}
      />
      <UpdateCountButton
        onClick={handleIncrement}
        text={INCREASE_BTN_TEXT}
        disabled={counterAsNumber >= max}
      />
    </div>
  );
};

export default SelectNumber;
