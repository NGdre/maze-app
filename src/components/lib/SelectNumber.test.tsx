import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SelectNumber from "./SelectNumber";

describe(`${SelectNumber.name} Component`, () => {
  const min = 50,
    max = 200,
    beetween = 100;

  it("should take a snapshot", () => {
    const { asFragment } = render(
      <SelectNumber
        onSelect={() => {}}
        value={beetween}
        min={min}
        max={max}
        labelContent="label"
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test(`renders initial count of ${beetween}`, () => {
    const expected = beetween;

    render(
      <SelectNumber
        onSelect={(number) => {
          expect(number).toBe(expected);
        }}
        value={expected}
        min={min}
        max={max}
      />
    );
    const countElement = screen.getByDisplayValue(expected);
    expect(countElement).toBeInTheDocument();
  });

  test("increments count when increment button is clicked", () => {
    const expected = beetween + 1;

    const onSelect = jest.fn((number) => number);

    render(
      <SelectNumber onSelect={onSelect} value={beetween} min={min} max={max} />
    );

    const incrementButton = screen.getByText(/\+/);
    fireEvent.click(incrementButton);
    const countElement = screen.getByDisplayValue(expected);

    expect(onSelect.mock.calls[1][0]).toBe(expected);
    expect(countElement).toBeInTheDocument();
  });

  test("decrements count when decrement button is clicked", () => {
    const expected = beetween - 1;

    const onSelect = jest.fn((number) => number);

    render(
      <SelectNumber onSelect={onSelect} value={beetween} min={min} max={max} />
    );

    const decrementButton = screen.getByText(/\-/);
    fireEvent.click(decrementButton);
    const countElement = screen.getByDisplayValue(expected);

    expect(onSelect.mock.calls[1][0]).toBe(expected);
    expect(countElement).toBeInTheDocument();
  });

  test("correctly updates count with multiple clicks", () => {
    const expected = beetween + 1;

    const onSelect = jest.fn((number) => number);

    render(
      <SelectNumber onSelect={onSelect} value={beetween} min={min} max={max} />
    );

    const incrementButton = screen.getByText(/\+/);
    const decrementButton = screen.getByText(/\-/);

    fireEvent.click(incrementButton);
    fireEvent.click(incrementButton);
    fireEvent.click(decrementButton);

    const countElement = screen.getByDisplayValue(expected);

    expect(onSelect.mock.calls[1][0]).toBe(expected);
    expect(countElement).toBeInTheDocument();
  });

  test("doesn't increment when count equals max value", () => {
    render(
      <SelectNumber
        onSelect={(number) => {
          expect(number).toBe(max);
        }}
        value={max}
        min={min}
        max={max}
      />
    );
    const incrementButton = screen.getByText(/\+/);
    fireEvent.click(incrementButton);
    const countElement = screen.getByDisplayValue(max);
    expect(countElement).toBeInTheDocument();
  });

  test("doesn't decrement when count equals min value", () => {
    render(
      <SelectNumber
        onSelect={(number) => {
          expect(number).toBe(min);
        }}
        value={min}
        min={min}
        max={max}
      />
    );
    const decrementButton = screen.getByText(/\-/);
    fireEvent.click(decrementButton);
    const countElement = screen.getByDisplayValue(min);
    expect(countElement).toBeInTheDocument();
  });

  test("produces a clamped value on input blur", async () => {
    const onSelect = jest.fn();
    render(
      <SelectNumber onSelect={onSelect} value={max} min={min} max={max} />
    );
    const input = screen.getByDisplayValue(max);
    const user = userEvent.setup();

    // Test value below min
    await user.clear(input);
    await user.type(input, "0");
    await user.tab(); // Triggers blur

    expect(onSelect).toHaveBeenLastCalledWith(min);
    expect(screen.getByDisplayValue(min)).toBeInTheDocument();

    // Test value above max
    await user.clear(input);
    await user.type(input, String(max + 1));
    await user.tab();

    expect(onSelect).toHaveBeenLastCalledWith(max);
    expect(screen.getByDisplayValue(max)).toBeInTheDocument();
  });

  test("calls onSelect only when typed correct value", async () => {
    const onSelect = jest.fn((number) => number);
    const defaultValue = beetween;

    render(
      <SelectNumber
        onSelect={onSelect}
        value={defaultValue}
        min={min}
        max={max}
      />
    );

    const input = screen.getByDisplayValue(defaultValue);

    const user = userEvent.setup();

    await user.clear(input);
    await user.type(input, "0");
    await user.keyboard("[Tab]");

    expect(onSelect.mock.calls[0][0]).toBe(defaultValue);
    expect(onSelect.mock.calls[1][0]).toBe(min);
  });
});
