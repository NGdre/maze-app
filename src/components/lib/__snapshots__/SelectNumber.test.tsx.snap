// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectNumber Component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-3"
  >
    <div
      class="text-center col-span-3"
    >
      <label
        class="text-gray-700 text-sm font-semibold capitalize"
        for=":r0:"
      >
        label
      </label>
    </div>
    <button
      class="bg-gray-300 text-gray-600 rounded-full outline-none cursor-pointer hover:text-white hover:bg-blue-500"
    >
      <span
        class="m-auto text-2xl"
      >
        -
      </span>
    </button>
    <input
      class="outline-none focus:outline-none text-center font-semibold 
              hover:text-gray-500 focus:text-black md:text-base flex items-center"
      id=":r0:"
      max="200"
      min="50"
      type="number"
      value="100"
    />
    <button
      class="bg-gray-300 text-gray-600 rounded-full outline-none cursor-pointer hover:text-white hover:bg-blue-500"
    >
      <span
        class="m-auto text-2xl"
      >
        +
      </span>
    </button>
  </div>
</DocumentFragment>
`;
